### PR TITLE
crafting valid doi from invalid uri

### PR DESCRIPTION
**Pull request by truhacevkir**

### CODE REVIEW

File: src/test/java/no/sikt/nva/scrapers/DublinCoreScraperTest.java [hunk 1]
```diff
@@ -790,6 +790,20 @@ var record = dcScraper.validateAndParseDublinCore(dublinCore, brageLocation, SOM
         assertThat(record.getLink(), is(nullValue()));
     }
 
+    @ParameterizedTest
+    @ValueSource(strings = {"https://doi.org/10.1577/1548-8667(1998)010<0056:EOOAFI>2.0.CO;2",
+        "https://doi.org/10.1890/0012-9658(2006)87[2915:DMWITM]2.0.CO;2",
+        "https://doi.org/10.2983/0730-8000(2008)27[525:EOAMRF]2.0.CO;2"})
+    void shouldCreateDoiWhenDoiIsNotValidUriButIsValisDoi(String value) {
```
Comment (truhacevkir) [inline src/test/java/no/sikt/nva/scrapers/DublinCoreScraperTest.java:797]: Creating doi's from invalid uri's

File: src/test/java/no/sikt/nva/validators/DoiValidatorTest.java [hunk 2]
```diff
@@ -60,4 +45,14 @@ void shouldReturnEmptyErrorListWhenDoiIsEmpty() {
         assertThat(actualDoiErrorsOnline, is(equalTo(expectedDoiErrors)));
     }
 
+    @ParameterizedTest
+    @MethodSource("doiProvider")
+    void shouldNotCreateErrorWhenDoiIsValid(String doi) {
```
Comment (truhacevkir) [inline src/test/java/no/sikt/nva/validators/DoiValidatorTest.java:50]: Not logging invalid uri's when they are valid doi's

File: src/main/java/no/sikt/nva/validators/DoiValidator.java [hunk 3]
```diff
@@ -156,7 +182,13 @@ private static List<String> extractDoiList(DublinCore dublinCore) {
     }
 
     private static boolean isValidDoiOffline(String doi) {
-        return nva.commons.doi.DoiValidator.validateOffline(doi);
+        return isValidDoi(doi);
+    }
+
+    public static boolean isValidDoi(String doi) {
+        var doiRegex = "^(https?://doi\\.org/)?10.\\d{4,9}/[-._;():A-Z0-9<>\\[\\]]+";
```
Comment (joachimjorgensen) [inline src/main/java/no/sikt/nva/validators/DoiValidator.java:189]: Where does this magic come from?

File: src/main/java/no/sikt/nva/validators/DoiValidator.java [hunk 4]
```diff
@@ -51,6 +56,23 @@ public static Optional<ArrayList<ErrorDetails>> getDoiErrorDetailsOffline(Dublin
...
+        } catch (URISyntaxException e) {
+            return valueWithEncodedLastPath(value);
+        }
+    }
+
+    private static String valueWithEncodedLastPath(String value) {
+        var parts = value.split(SLASH);
+        var lastPath = parts[parts.length - 1];
+        var encodedLastPath = URLEncoder.encode(lastPath, StandardCharsets.UTF_8);
+        parts[parts.length - 1] = encodedLastPath;
```
Comment (brinxmat) [inline src/main/java/no/sikt/nva/validators/DoiValidator.java:72]: probably use substring and lastIndexOf

**Review by joachimjorgensen — APPROVED** [review]

File: src/main/java/no/sikt/nva/validators/DoiValidator.java (hunk 4 above)
Comment (brinxmat) [reply src/main/java/no/sikt/nva/validators/DoiValidator.java:72]: Also, test with trailing slash

File: src/main/java/no/sikt/nva/validators/DoiValidator.java [hunk 5]
```diff
@@ -59,7 +81,7 @@ public static String updateDoiStructureIfNeeded(String inputDoi) {
         if (doi.contains(DOI_DOMAIN_NAME)) {
             return HTTPS_STRING + doi;
         }
-        if (doi.contains(COLON)) {
+        if (doi.contains(COLON_V1) || doi.contains(COLON_V2)) {
```
Comment (brinxmat) [inline src/main/java/no/sikt/nva/validators/DoiValidator.java:84]: doi.toLowerCase

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: src/main/java/no/sikt/nva/validators/DoiValidator.java (hunk 3 above)
Comment (truhacevkir) [reply src/main/java/no/sikt/nva/validators/DoiValidator.java:189]: It is regex that allows doi's of all versions (I think so). It validates correctly invalid doi's and passes "correct" doi's.

**Review by anetteOlli — APPROVED** [review]
Bra jobba, det var sikkert ganske smerte å jobbe med dette.

**Pull request merged**